server/statshouse/worker-stats-buffer.cpp

```diff
@@ -0,0 +1,124 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2022 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#include "worker-stats-buffer.h"
+#include "common/tl/constants/statshouse.h"
+#include "runtime/critical_section.h"
+#include "statshouse-client.h"
+#include "common/resolver.h"
+
+namespace statshouse {
+
+constexpr size_t buffer_size = 50000;
+static char buffer[buffer_size];
+
+bool StatsBuffer::is_need_to_flush() {
+  return size >= data.size();
+}
+
+bool StatsBuffer::empty() const {
+  return size == 0;
+}
+
+void StatsBuffer::add_stat(double value) {
+  data[size++] = value;
+}
+
+std::vector<double> StatsBuffer::get_data_and_reset_buffer() {
+  std::vector<double> result(data.begin(), data.begin() + size);
+  size = 0;
+  return result;
+}
+
+WorkerStatsBuffer::WorkerStatsBuffer() {
+  last_send_time = std::chrono::steady_clock::now();
+}
+
+void WorkerStatsBuffer::add_query_stat(GenericQueryStatKey key, WorkerType worker_type, double value) {
+  auto &worker_type_buffer = generic_query_stats[static_cast<size_t>(worker_type)];
+  auto &stats_buffer = worker_type_buffer[static_cast<size_t>(key)];
+  auto now = std::chrono::steady_clock::now();
+
+  if (stats_buffer.is_need_to_flush() || now - last_send_time >= std::chrono::seconds{1}) {
+    flush();
+  }
+  stats_buffer.add_stat(value);
+}
+
+void WorkerStatsBuffer::add_query_stat(QueryStatKey key, double value) {
+  auto &stats_buffer = query_stats[static_cast<size_t>(key)];
+  auto now = std::chrono::steady_clock::now();
+
+  if (stats_buffer.is_need_to_flush() || now - last_send_time >= std::chrono::seconds{1}) {
+    last_send_time = now;
+    flush();
+  }
+  stats_buffer.add_stat(value);
+}
+
+void WorkerStatsBuffer::make_generic_metric(std::vector<StatsHouseMetric> &metrics, const char *name, GenericQueryStatKey stat_key, size_t worker_type,
+                                            const std::vector<tag> &tags) {
+  auto &stats_buffer = generic_query_stats[worker_type][static_cast<size_t>(stat_key)];
+  if (!stats_buffer.empty()) {
+    metrics.push_back(make_statshouse_value_metrics(name, stats_buffer.get_data_and_reset_buffer(), tags));
```

Comment:
You forgot to update `last_send_time` here, let's update it in `flush` only

Reply:
done